src/client/utils/property-overriding.ts

```diff
@@ -54,3 +55,56 @@ export function overrideDescriptor<O extends object, K extends keyof O> (obj: O,
 
     nativeMethods.objectDefineProperty(obj, prop, descriptor);
 }
+
+// NOTE: saves the original string representation of a native function in its wrapper as an additional property
+export function overrideStringRepresentation (nativeFnWrapper: Function, nativeFn: Function): void {
+    // NOTE: we need to override the 'name' property of our wrapper 
+    // to be the same as the `name` property of a native function
+    nativeMethods.objectDefineProperty(nativeFnWrapper, 'name', {
+        value: nativeFn['name']
+    });
+    
+    const nativeStringRepresentation = nativeMethods.Function.prototype.toString.call(nativeFn);
+
+    nativeMethods.objectDefineProperty(nativeFnWrapper, INTERNAL_PROPS.nativeStringRepresentation, {
+        value: nativeStringRepresentation,
+        configurable: true
+    });
+}
+
+export function isNativeFunction (fn: Function): boolean {
+    return !nativeMethods.objectHasOwnProperty.call(fn, INTERNAL_PROPS.nativeStringRepresentation);
+}
+
+export function overrideFunction<O extends object, K extends keyof O> (obj: O, fnName: K, wrapper: Function): void {
+    const fn = obj[fnName] as unknown as Function;
+
+    if (isNativeFunction(fn)) {
+        overrideStringRepresentation(wrapper, fn);
+        
+        nativeMethods.objectDefineProperty(obj, fnName, {
+            value: wrapper
+        });
+    }
+}
+
+export function overrideConstructor<O extends object, K extends keyof O> (obj: O, fnName: K, wrapper: Function, overrideProtoConstructor: boolean = false): void {
+    const nativePrototype = obj[fnName]['prototype'];
+    
+    overrideFunction(obj, fnName, wrapper);
+
+    if (nativePrototype) {
```

Comment:
I don't understand why we need this condition. Each function has the `prototype` property and that `prototype` has `constructor`.